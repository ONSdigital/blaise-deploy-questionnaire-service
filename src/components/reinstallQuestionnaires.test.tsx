/**
 * @jest-environment jsdom
 */

import React from "react";
import { cleanup, fireEvent, render, screen, waitFor } from "@testing-library/react";
import "@testing-library/jest-dom";
import flushPromises from "../tests/utils";
import { act } from "react-dom/test-utils";
import { createMemoryHistory } from "history";
import { Router } from "react-router";
import { Questionnaire } from "blaise-api-node-client";
import ReinstallQuestionnaires from "./reinstallQuestionnaires";
import axios from "axios";
import MockAdapter from "axios-mock-adapter";

const mock = new MockAdapter(axios);

const questionnaireList: Questionnaire[] = [{
    name: "OPN2101A",
    serverParkName: "gusty",
    installDate: "2021-01-15T14:41:29.4399898+00:00",
    status: "Active",
    dataRecordCount: 0,
    hasData: false,
    active: false,
    fieldPeriod: "January 2021"
}];

const bucketQuestionnaire: string[] = ["OPN2101A.bpkg", "OPN2004A.bpkg", "LMS2101_BK2.bpkg"];

describe("Reinstall questionnaires list", () => {
    beforeEach(() => {
        mock.onGet("/api/questionnaires").reply(200, questionnaireList);
        mock.onGet("/bucket/files").reply(200, bucketQuestionnaire);
    });

    afterEach(() => {
        jest.clearAllMocks();
        cleanup();
        mock.reset();
    });

    it("view Blaise Status page matches Snapshot", async () => {
        const history = createMemoryHistory();
        const wrapper = render(
            <Router history={history}>
                <ReinstallQuestionnaires />
            </Router>
        );

        await act(async () => {
            await flushPromises();
        });

        await waitFor(() => {
            expect(wrapper).toMatchSnapshot();
        });
    });

    it("should render correctly", async () => {
        const history = createMemoryHistory();
        render(
            <Router history={history}>
                <ReinstallQuestionnaires />
            </Router>
        );

        await act(async () => {
            await flushPromises();
            await flushPromises();
        });

        await waitFor(() => {
            expect(screen.getByText(/Reinstall questionnaire/i)).toBeDefined();
            expect(screen.getByText(/OPN2004A/i)).toBeDefined();
            expect(screen.getByText(/LMS2101_BK2/i)).toBeDefined();
            expect(screen.queryByText(/OPN2101A/i)).not.toBeInTheDocument();
        });
    });

    it("should render a message if all items in bucket are already installed", async () => {
        const questionnaireList: any[] = [{ name: "OPN2101A" }, { name: "OPN2004A" }, { name: "LMS2101_BK2" }];

        mock.onGet("/api/questionnaires").reply(200, questionnaireList);

        const history = createMemoryHistory();
        render(
            <Router history={history}>
                <ReinstallQuestionnaires />
            </Router>
        );

        await act(async () => {
            await flushPromises();
        });

        await waitFor(() => {
            expect(screen.getByText(/Reinstall questionnaire/i)).toBeDefined();
            expect(screen.getByText(/No compatible previously installed questionnaires found./i)).toBeDefined();
        });
    });
});

describe("Reinstall questionnaires", () => {
    beforeEach(() => {
        mock.onGet("/api/questionnaires").reply(200, []);
        mock.onGet("/bucket/files").reply(200, bucketQuestionnaire);
        mock.onGet("/upload/verify?filename=OPN2004A.bpkg").reply(200, { name: "OPN2004A.bpkg" });
        mock.onPost("/api/install").reply(201);
    });

    afterEach(() => {
        jest.clearAllMocks();
        cleanup();
        mock.reset();
    });

    it("should redirect to the success page after install", async () => {
        const history = createMemoryHistory();
        render(
            <Router history={history}>
                <ReinstallQuestionnaires />
            </Router>
        );

        await act(async () => {
            await flushPromises();
        });

        await act(async () => {
            fireEvent.click(screen.getByText(/OPN2004A/i));
            await flushPromises();
            fireEvent.click(screen.getByText(/Install selected questionnaire/i));
            await flushPromises();
        });

        await waitFor(() => {
            // Check page has been redirected to summary page
            expect(history.location.pathname).toEqual("/UploadSummary");
            // State should be blank as its successful
            expect(history.location.state).toEqual({ questionnaireName: "OPN2004A", status: "" });
        });
    });

    it("should render a message if all items in bucket are already installed", async () => {
        const questionnaireList: any[] = [{ name: "OPN2101A" }, { name: "OPN2004A" }, { name: "LMS2101_BK2" }];

        mock.onGet("/api/questionnaires").reply(200, questionnaireList);

        const history = createMemoryHistory();
        render(
            <Router history={history}>
                <ReinstallQuestionnaires />
            </Router>
        );

        await act(async () => {
            await flushPromises();
        });

        await waitFor(() => {
            expect(screen.getByText(/Reinstall questionnaire/i)).toBeDefined();
            expect(screen.getByText(/No compatible previously installed questionnaires found./i)).toBeDefined();
        });
    });
});

describe("Given the API returns a 500 status", () => {
    beforeEach(() => {
        mock.onGet("/api/questionnaires").reply(500);
    });

    afterEach(() => {
        jest.clearAllMocks();
        cleanup();
        mock.reset();
    });

    it("it should render with the error message displayed", async () => {
        const history = createMemoryHistory();
        render(
            <Router history={history}>
                <ReinstallQuestionnaires />
            </Router>
        );

        await waitFor(() => {
            expect(screen.getByText(/Unable to load questionnaires./i)).toBeDefined();
            expect(screen.queryByText(/Loading/i)).not.toBeInTheDocument();
        });
    });
});

describe("Given the API returns an empty list", () => {
    beforeEach(() => {
        mock.onGet("/api/questionnaires").reply(200, []);
        mock.onGet("/bucket/files").reply(200, []);
    });

    afterEach(() => {
        jest.clearAllMocks();
        cleanup();
        mock.reset();
    });

    it("it should render with a message to inform the user in the list", async () => {
        const history = createMemoryHistory();
        render(
            <Router history={history}>
                <ReinstallQuestionnaires />
            </Router>
        );

        await act(async () => {
            await flushPromises();
        });

        await waitFor(() => {
            expect(screen.getByText(/Reinstall questionnaire/i)).toBeDefined();
            expect(screen.getByText(/No compatible previously installed questionnaires found./i)).toBeDefined();
        });
    });
});
