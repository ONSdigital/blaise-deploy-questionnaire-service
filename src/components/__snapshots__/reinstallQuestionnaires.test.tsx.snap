// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Reinstall questionnaires list view Blaise Status page matches Snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <nav
        aria-label="Breadcrumb"
        class="ons-breadcrumb"
      >
        <ol
          class="ons-breadcrumb__items u-fs-s"
        >
          <li
            class="ons-breadcrumb__item"
            id="breadcrumb-0"
          >
            <a
              class="ons-breadcrumb__link"
              href="/"
            >
              Home
            </a>
            <svg
              class="ons-svg-icon"
              focusable="false"
              viewBox="0 0 8 13"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M5.74,14.28l-.57-.56a.5.5,0,0,1,0-.71h0l5-5-5-5a.5.5,0,0,1,0-.71h0l.57-.56a.5.5,0,0,1,.71,0h0l5.93,5.93a.5.5,0,0,1,0,.7L6.45,14.28a.5.5,0,0,1-.71,0Z"
                transform="translate(-5.02 -1.59)"
              />
            </svg>
          </li>
          <li
            class="ons-breadcrumb__item"
            id="breadcrumb-1"
          >
            <a
              class="ons-breadcrumb__link"
              href="/upload"
            >
              Deploy a questionnaire
            </a>
            <svg
              class="ons-svg-icon"
              focusable="false"
              viewBox="0 0 8 13"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M5.74,14.28l-.57-.56a.5.5,0,0,1,0-.71h0l5-5-5-5a.5.5,0,0,1,0-.71h0l.57-.56a.5.5,0,0,1,.71,0h0l5.93,5.93a.5.5,0,0,1,0,.7L6.45,14.28a.5.5,0,0,1-.71,0Z"
                transform="translate(-5.02 -1.59)"
              />
            </svg>
          </li>
        </ol>
      </nav>
      <main
        class="ons-page__main ons-u-mt-no"
        id="main-content"
      >
        <h1
          class="ons-u-mb-l"
        >
          Reinstall questionnaire
        </h1>
        <p>
          Reinstall a previously uploaded questionnaire.
          <br />
          This will always deploy the last uploaded version of the questionnaire.
        </p>
        <div
          class="elementToFadeIn"
        >
          <form>
            <fieldset
              class="ons-fieldset"
            >
              <legend
                class="ons-fieldset__legend"
              >
                Select a questionnaire to install
              </legend>
              <div
                class="ons-radios__items"
              >
                <p
                  class="ons-radios__item"
                >
                  <span
                    class="ons-radio"
                  >
                    <input
                      aria-label="No"
                      class="ons-radio__input ons-js-radio "
                      id="install-OPN2004A.bpkg"
                      name="select-survey"
                      type="radio"
                      value="OPN2004A.bpkg"
                    />
                    <label
                      class="ons-radio__label "
                      for="install-OPN2004A.bpkg"
                    >
                      OPN2004A.bpkg
                    </label>
                  </span>
                </p>
                <br />
                <p
                  class="ons-radios__item"
                >
                  <span
                    class="ons-radio"
                  >
                    <input
                      aria-label="No"
                      class="ons-radio__input ons-js-radio "
                      id="install-LMS2101_BK2.bpkg"
                      name="select-survey"
                      type="radio"
                      value="LMS2101_BK2.bpkg"
                    />
                    <label
                      class="ons-radio__label "
                      for="install-LMS2101_BK2.bpkg"
                    >
                      LMS2101_BK2.bpkg
                    </label>
                  </span>
                </p>
                <br />
              </div>
            </fieldset>
            <br />
            <button
              class="ons-btn "
              data-testid="button"
              id="confirm-install"
              type="button"
            >
              <span
                class="ons-btn__inner"
              >
                Install selected questionnaire
              </span>
            </button>
          </form>
        </div>
      </main>
    </div>
  </body>,
  "container": <div>
    <nav
      aria-label="Breadcrumb"
      class="ons-breadcrumb"
    >
      <ol
        class="ons-breadcrumb__items u-fs-s"
      >
        <li
          class="ons-breadcrumb__item"
          id="breadcrumb-0"
        >
          <a
            class="ons-breadcrumb__link"
            href="/"
          >
            Home
          </a>
          <svg
            class="ons-svg-icon"
            focusable="false"
            viewBox="0 0 8 13"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M5.74,14.28l-.57-.56a.5.5,0,0,1,0-.71h0l5-5-5-5a.5.5,0,0,1,0-.71h0l.57-.56a.5.5,0,0,1,.71,0h0l5.93,5.93a.5.5,0,0,1,0,.7L6.45,14.28a.5.5,0,0,1-.71,0Z"
              transform="translate(-5.02 -1.59)"
            />
          </svg>
        </li>
        <li
          class="ons-breadcrumb__item"
          id="breadcrumb-1"
        >
          <a
            class="ons-breadcrumb__link"
            href="/upload"
          >
            Deploy a questionnaire
          </a>
          <svg
            class="ons-svg-icon"
            focusable="false"
            viewBox="0 0 8 13"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M5.74,14.28l-.57-.56a.5.5,0,0,1,0-.71h0l5-5-5-5a.5.5,0,0,1,0-.71h0l.57-.56a.5.5,0,0,1,.71,0h0l5.93,5.93a.5.5,0,0,1,0,.7L6.45,14.28a.5.5,0,0,1-.71,0Z"
              transform="translate(-5.02 -1.59)"
            />
          </svg>
        </li>
      </ol>
    </nav>
    <main
      class="ons-page__main ons-u-mt-no"
      id="main-content"
    >
      <h1
        class="ons-u-mb-l"
      >
        Reinstall questionnaire
      </h1>
      <p>
        Reinstall a previously uploaded questionnaire.
        <br />
        This will always deploy the last uploaded version of the questionnaire.
      </p>
      <div
        class="elementToFadeIn"
      >
        <form>
          <fieldset
            class="ons-fieldset"
          >
            <legend
              class="ons-fieldset__legend"
            >
              Select a questionnaire to install
            </legend>
            <div
              class="ons-radios__items"
            >
              <p
                class="ons-radios__item"
              >
                <span
                  class="ons-radio"
                >
                  <input
                    aria-label="No"
                    class="ons-radio__input ons-js-radio "
                    id="install-OPN2004A.bpkg"
                    name="select-survey"
                    type="radio"
                    value="OPN2004A.bpkg"
                  />
                  <label
                    class="ons-radio__label "
                    for="install-OPN2004A.bpkg"
                  >
                    OPN2004A.bpkg
                  </label>
                </span>
              </p>
              <br />
              <p
                class="ons-radios__item"
              >
                <span
                  class="ons-radio"
                >
                  <input
                    aria-label="No"
                    class="ons-radio__input ons-js-radio "
                    id="install-LMS2101_BK2.bpkg"
                    name="select-survey"
                    type="radio"
                    value="LMS2101_BK2.bpkg"
                  />
                  <label
                    class="ons-radio__label "
                    for="install-LMS2101_BK2.bpkg"
                  >
                    LMS2101_BK2.bpkg
                  </label>
                </span>
              </p>
              <br />
            </div>
          </fieldset>
          <br />
          <button
            class="ons-btn "
            data-testid="button"
            id="confirm-install"
            type="button"
          >
            <span
              class="ons-btn__inner"
            >
              Install selected questionnaire
            </span>
          </button>
        </form>
      </div>
    </main>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
